src/cdomains/valueDomain.ml

```diff
@@ -108,21 +107,23 @@ struct
     | TNamed ({tname = "pthread_t"; _}, _) -> true
     | _ -> false
 
-  let rec bot_value (t: typ): t =
+  let rec bot_value ?(varAttr=[]) (t: typ): t =
     match t with
     | _ when is_mutex_type t -> `Mutex
     | TInt _ -> `Bot (*`Int (ID.bot ()) -- should be lower than any int or address*)
     | TFloat _ -> `Bot
     | TPtr _ -> `Address (AD.bot ())
-    | TComp ({cstruct=true; _} as ci,_) -> `Struct (Structs.create (fun fd -> bot_value fd.ftype) ci)
+    | TComp ({cstruct=true; _} as ci,_) -> `Struct (Structs.create (fun fd -> bot_value ~varAttr:fd.fattr fd.ftype) ci)
     | TComp ({cstruct=false; _},_) -> `Union (Unions.bot ())
     | TArray (ai, None, _) ->
-      `Array (CArrays.make (IndexDomain.bot ()) (bot_value ai))
+      let typAttr = typeAttrs ai in
+      `Array (CArrays.make ~varAttr ~typAttr (IndexDomain.bot ()) (bot_value ai))
     | TArray (ai, Some exp, _) ->
+      let typAttr = typeAttrs ai in
       let l = BatOption.map Cilint.big_int_of_cilint (Cil.getInteger (Cil.constFold true exp)) in
-      `Array (CArrays.make (BatOption.map_default (IndexDomain.of_int (Cilfacade.ptrdiff_ikind ())) (IndexDomain.bot ()) l) (bot_value ai))
+      `Array (CArrays.make ~varAttr ~typAttr (BatOption.map_default (IndexDomain.of_int (Cilfacade.ptrdiff_ikind ())) (IndexDomain.bot ()) l) (bot_value ai))
     | t when is_thread_type t -> `Thread (ConcDomain.ThreadSet.empty ())
-    | TNamed ({ttype=t; _}, _) -> bot_value t
+    | TNamed ({ttype=t; _}, _) -> bot_value ~varAttr (unrollType t)
```

Comment:
Why is there a `unrollType` here now? This way the outermost `TNamed` is handled here, but then the inner ones are removed by `unrollType`, which seems unnecessary since `bot_value` does that recursion anyway without having to allocate an intermediate unrolled type for it.

Reply:
unrollType also collects the attributes of the named types, which is why I changed it here

Reply:
I think those could also be collected here, right now they're just ignored with wildcard (`_`).

This is in `bot_value`, but other functions like `top_value`/`init_value` have other similar changes, but without the `unrollType`. Shouldn't this attribute collecting be consistent everywhere?

Reply:
I thought it was simpler to use unrollType than to add another list parameter for manually collecting the type attributes.

It should be the same in the other functions, I just reverted it to the wrong version after trying something out.

Reply:
When collecting the attributes manually with the code below, I get an Out of memory exception:

```
  let rec bot_value ?(varAttr=[]) ?(typAttr=[]) (t: typ): t =
    match t with
    [...]
    | TArray (ai, None, _) ->
      let typAttr = addAttributes (typeAttrs ai) typAttr in
      `Array (CArrays.make ~varAttr ~typAttr (IndexDomain.bot ()) (bot_value ai))
    [...]
    | TNamed ({ttype=t; _}, attr) -> bot_value ~varAttr ~typAttr:(addAttributes attr typAttr) t
```
I do not understand why this runs out of memory so quickly and the stats do not match. 
```
Memory statistics: total=23.51MB, max=6.63MB, minor=21.73MB, major=5.82MB, promoted=4.04MB
    minor collections=10  major collections=1 compactions=0


Fatal error: exception Out of memory
```

Reply:
Right now the CI shows out of memory exceptions for many tests even without that change, still with that `unrollType`. So there seems to be some other issue.

When running with `-v`, the full stacktrace of the error is shown, which starts with:
```
Fatal error: exception Out of memory
Raised by primitive operation at BatHashtbl.find_option.loop in file "src/batHashtbl.mlv", line 218, characters 9-16
Called from Goblint_lib__Base.MainFunctor.add_to_array_map in file "src/analyses/base.ml", line 110, characters 10-51
```
If I were to guess, the problem is the use of the polymorphic `Hashtbl` instead of `VarH` as suggested in another comment. Polymorphic hashtables use polymorphic equality `(=)` to compare elements, but that's incorrect on many CIL types, including `varinfo`. Since CIL data structures contain cyclic references, the default polymorphic equality will recurse forever and very quickly run out of memory (in the C runtime, which is probably why it doesn't show in GC stats). Using a hashtable with our custom equality that avoids the issue should fix the problem.